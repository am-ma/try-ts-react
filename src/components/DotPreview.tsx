import React, { CSSProperties } from 'react';
import DotValues from '~/models/DotValues.model';

interface DotPreviewProps {
  dots: DotValues[];
  dotSize: number;
  height: number;
  width: number;
}

interface DotPreviewState {
  dlUrl: string;
}

export default class DotPreview extends React.Component<DotPreviewProps, DotPreviewState> {
  private svg = React.createRef<SVGSVGElement>();
  private canvas = React.createRef<HTMLCanvasElement>();
  constructor(props: DotPreviewProps) {
    super(props);
    this.state = {
      dlUrl: '#',
    };
  }
  render() {
    // render svg
    const dotSize = this.props.dotSize;
    const width = dotSize * this.props.width;
    const height = dotSize * this.props.height;

    const imageDots = this.props.dots
      .map((dot, i) => {
        if (!dot.isFill) {
          return null;
        }
        const style: CSSProperties = {
          fill: dot.color,
        };
        const xIndex = i % this.props.height;
        const yIndex = Math.floor(i / this.props.width);
        return (
          <rect
            key={`${xIndex}_${yIndex}`}
            x={xIndex * dotSize}
            y={yIndex * dotSize}
            width={dotSize}
            height={dotSize}
            style={style}></rect>
        );
      })
      .filter(dotJSX => dotJSX !== null);

    return (
      <>
        <svg id="svg" width={width} height={height} ref={this.svg}>
          {imageDots}
        </svg>
        <canvas id="canvas" ref={this.canvas} width={width} height={height} className="not-display"></canvas>
        <div>
          <a href={this.state.dlUrl} id="download" download="dots.png">
            download
          </a>
        </div>
      </>
    );
  }

  componentDidMount() {
    this.renderForDownload();
  }

  componentDidUpdate() {
    this.renderForDownload();
  }

  renderForDownload() {
    // svg to xml data src
    const svg = this.svg.current!;
    const svgData = new XMLSerializer().serializeToString(svg as Node);
    const imgsrc = 'data:image/svg+xml;charset=utf-8;base64,' + btoa(unescape(encodeURIComponent(svgData)));

    const canvas = this.canvas.current!;
    const ctx = canvas.getContext('2d') as CanvasRenderingContext2D;
    const image = new Image();
    image.onload = () => {
      // render canvas
      ctx.drawImage(image, 0, 0);
      // render download button
      this.setState({
        dlUrl: canvas.toDataURL('image/png'),
      });
    };
    image.src = imgsrc;
  }
}
