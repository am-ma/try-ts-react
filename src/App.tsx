import React from 'react';

import DotValues from '~/models/DotValues.model';
import DotCanvas from './components/DotCanvas';
import DotPreview from './components/DotPreview';

interface AppProps {
  canvasHeight: number;
  canvasWidth: number;
}
interface AppState {
  color: string;
  dots: DotValues[];
  isSaved: boolean;
  dotSize: number;
}

export default class App extends React.Component<AppProps, AppState> {
  private defaultColor = '#000';
  constructor(props: AppProps) {
    super(props);

    // 読み出し
    const loaded = this.load();
    const isLoaded = loaded.length > 0;

    this.state = {
      color: this.defaultColor,
      dots: isLoaded ? loaded : this.createDefaultDots(),
      isSaved: isLoaded,
      dotSize: 8,
    };
  }

  createDefaultDots() {
    return Array(this.props.canvasHeight * this.props.canvasWidth)
      .fill(null)
      .map(
        () =>
          ({
            isFill: false,
            color: this.defaultColor,
          } as DotValues)
      );
  }

  handleClick(i: number) {
    const dots = this.state.dots.slice();
    dots[i].isFill = !dots[i].isFill;
    dots[i].color = this.state.color;
    this.setState({ dots: dots });
  }

  colorOnChange(e: React.ChangeEvent<HTMLInputElement>) {
    this.setState({
      color: e.target.value,
    });
  }

  clearOnClick() {
    const dots = this.createDefaultDots();
    this.setState({
      dots,
    });
  }

  saveOnClick() {
    const dotsJson = JSON.stringify(this.state.dots);
    window.localStorage.setItem('ts-react-drawing', dotsJson);
    window.alert('saved!');
    this.setState({ isSaved: true });
  }

  dotSizeOnChange(e: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ dotSize: parseInt(e.target.value) });
  }

  load() {
    const dotsJson = window.localStorage.getItem('ts-react-drawing');
    if (!dotsJson) {
      return [];
    }

    return JSON.parse(dotsJson);
  }

  loadOnClick() {
    const loaded = this.load();
    this.setState({ dots: loaded });
  }

  render() {
    return (
      <div className="board">
        <div className="board-board">
          <DotCanvas
            height={this.props.canvasHeight}
            width={this.props.canvasWidth}
            dots={this.state.dots}
            onClick={(i: number) => this.handleClick(i)}
          />
        </div>
        <div className="board-info">
          <div>
            <input
              type="color"
              value={this.state.color}
              onChange={(e: React.ChangeEvent<HTMLInputElement>) => this.colorOnChange(e)}
            />
          </div>
          <div>
            <button type="button" onClick={() => this.clearOnClick()}>
              clear
            </button>
          </div>
          <div>
            <button type="button" onClick={() => this.saveOnClick()}>
              save
            </button>
            {this.state.isSaved && (
              <button type="button" onClick={() => this.loadOnClick()}>
                load
              </button>
            )}
          </div>
          <div className="downloads">
            <div>
              dot size{' '}
              <input
                className="dot-size"
                type="number"
                value={this.state.dotSize}
                onChange={e => this.dotSizeOnChange(e)}></input>{' '}
              px
            </div>
            <DotPreview
              height={this.props.canvasHeight}
              width={this.props.canvasWidth}
              dotSize={this.state.dotSize}
              dots={this.state.dots}
            />
          </div>
        </div>
      </div>
    );
  }
}
