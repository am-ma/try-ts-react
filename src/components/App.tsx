import React from 'react';

import DotValues from '~/models/DotValues.model';
import DotCanvas from './DotCanvas';
import DotPreview from './DotPreview';

const canvasHeight: number = 12;
const canvasWidth: number = 12;

interface AppProps {}
interface AppState {
  color: string;
  dots: DotValues[];
  isSaved: boolean;
  dotSize: number;
}

export default class App extends React.Component<AppProps, AppState> {
  constructor(props: AppProps) {
    super(props);
    const defaultColor = '#000';
    // 読み出し
    const loaded = this.load();

    this.state = {
      color: defaultColor,
      dots: loaded
        ? loaded
        : Array(canvasHeight * canvasWidth)
            .fill(null)
            .map(
              () =>
                ({
                  isFill: false,
                  color: defaultColor,
                } as DotValues)
            ),
      isSaved: loaded !== [],
      dotSize: 8,
    };
  }

  handleClick(i: number) {
    const dots = this.state.dots.slice();
    dots[i].isFill = !dots[i].isFill;
    dots[i].color = this.state.color;
    this.setState({ dots: dots });
  }

  colorOnChange(e: React.ChangeEvent<HTMLInputElement>) {
    this.setState({
      color: e.target.value,
    });
  }

  saveOnClick() {
    const dotsJson = JSON.stringify(this.state.dots);
    window.localStorage.setItem('ts-react-drawing', dotsJson);
    window.alert('saved!');
    this.setState({ isSaved: true });
  }

  dotSizeOnChange(e: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ dotSize: parseInt(e.target.value) });
  }

  load() {
    const dotsJson = window.localStorage.getItem('ts-react-drawing');
    if (!dotsJson) {
      return [];
    }

    return JSON.parse(dotsJson);
  }

  loadOnClick() {
    const loaded = this.load();
    this.setState({ dots: loaded });
  }

  render() {
    return (
      <div className="board">
        <div className="board-board">
          <DotCanvas
            height={canvasHeight}
            width={canvasWidth}
            dots={this.state.dots}
            onClick={(i: number) => this.handleClick(i)}
          />
        </div>
        <div className="board-info">
          <div>
            <input
              type="color"
              value={this.state.color}
              onChange={(e: React.ChangeEvent<HTMLInputElement>) => this.colorOnChange(e)}
            />
          </div>
          <div>
            <button type="button" onClick={() => this.saveOnClick()}>
              save
            </button>
            {this.state.isSaved && (
              <button type="button" onClick={() => this.loadOnClick()}>
                load
              </button>
            )}
          </div>
          <div className="downloads">
            <div>
              dot size{' '}
              <input
                className="dot-size"
                type="number"
                value={this.state.dotSize}
                onChange={e => this.dotSizeOnChange(e)}></input>{' '}
              px
            </div>
            <DotPreview height={canvasHeight} width={canvasWidth} dotSize={this.state.dotSize} dots={this.state.dots} />
          </div>
        </div>
      </div>
    );
  }
}
