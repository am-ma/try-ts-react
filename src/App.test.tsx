import React from 'react';
import App from './App';
import renderer from 'react-test-renderer';
import DotPreview from './components/DotPreview';
import { shallow } from 'enzyme';

describe('App', () => {
  const jsx = <App canvasHeight={2} canvasWidth={2} />;
  describe('render', () => {
    it('will render', () => {
      jest.spyOn(DotPreview.prototype, 'renderForDownload').mockImplementationOnce(jest.fn());
      const result = renderer.create(jsx).toJSON();

      expect(result).toMatchSnapshot();
    });
  });

  describe('handleClick', () => {
    it('will update target index Dot.values', () => {
      const wrapper = shallow(jsx);
      const app = wrapper.instance() as App;
      app.state.dots[0] = { isFill: false, color: '#000' };

      app.setState({ color: '#fff' });
      app.handleClick(0);
      expect(app.state.dots[0]).toEqual({ isFill: true, color: '#fff' });
    });
  });

  describe('colorOnChange', () => {
    it('will update color', () => {
      const wrapper = shallow(jsx);
      const app = wrapper.instance() as App;
      app.setState({ color: '#000' });

      const event: any = { target: { value: '#fff' } };
      app.colorOnChange(event);
      expect(app.state.color).toEqual('#fff');
    });
  });

  describe('saveOnClick', () => {
    it('will save dots', () => {
      const wrapper = shallow(jsx);
      const app = wrapper.instance() as App;

      const setItem = jest.fn();
      jest.spyOn(window.localStorage, 'setItem').mockImplementationOnce(setItem);
      const alert = jest.fn();
      jest.spyOn(window, 'alert').mockImplementationOnce(alert);

      app.setState({
        isSaved: false,
        dots: [],
      });

      app.saveOnClick();

      expect(setItem).toHaveBeenCalledWith('ts-react-drawing', '[]');
      expect(alert).toHaveBeenCalledWith('saved!');
      expect(app.state.isSaved).toBe(true);
    });
  });

  describe('dotSizeOnChange', () => {
    it('will save dotSize', () => {
      const event: any = { target: { value: 10 } };
      const wrapper = shallow(jsx);
      const app = wrapper.instance() as App;
      app.setState({
        dotSize: 2,
      });

      app.dotSizeOnChange(event);

      expect(app.state.dotSize).toBe(10);
    });
  });

  describe('load', () => {
    it('will load dots in localStorage', () => {
      window.localStorage.setItem('ts-react-drawing', '[{"isFill":false,"color":"#ccc"}]');

      const wrapper = shallow(jsx);
      const app = wrapper.instance() as App;

      const result = app.load();

      expect(result).toEqual([{ isFill: false, color: '#ccc' }]);
    });
    it('will return empty array when localStrage target item is empty', () => {
      window.localStorage.setItem('ts-react-drawing', '');

      const wrapper = shallow(jsx);
      const app = wrapper.instance() as App;

      const result = app.load();

      expect(result).toEqual([]);
    });
    it('will return empty array when localStrage target item is no set', () => {
      window.localStorage.clear();

      const wrapper = shallow(jsx);
      const app = wrapper.instance() as App;

      const result = app.load();

      expect(result).toEqual([]);
    });
  });

  describe('loadOnClick', () => {
    it('will update state.dots', () => {
      const wrapper = shallow(jsx);
      const app = wrapper.instance() as App;
      app.setState({
        dots: [],
      });
      jest.spyOn(app, 'load').mockImplementationOnce(() => ['hoge']);

      app.loadOnClick();

      expect(app.state.dots).toEqual(['hoge']);
    });
  });
});
