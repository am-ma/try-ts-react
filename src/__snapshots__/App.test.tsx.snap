// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render will render 1`] = `
<div
  className="board"
>
  <div
    className="board-board"
  >
    <div>
      <div
        className="canvas-row"
      >
        <button
          className="dot"
          onClick={[Function]}
          style={Object {}}
        />
        <button
          className="dot"
          onClick={[Function]}
          style={Object {}}
        />
      </div>
      <div
        className="canvas-row"
      >
        <button
          className="dot"
          onClick={[Function]}
          style={Object {}}
        />
        <button
          className="dot"
          onClick={[Function]}
          style={Object {}}
        />
      </div>
    </div>
  </div>
  <div
    className="board-info"
  >
    <div>
      <input
        onChange={[Function]}
        type="color"
        value="#000"
      />
    </div>
    <div>
      <button
        onClick={[Function]}
        type="button"
      >
        save
      </button>
    </div>
    <div
      className="downloads"
    >
      <div>
        dot size
         
        <input
          className="dot-size"
          onChange={[Function]}
          type="number"
          value={8}
        />
         
        px
      </div>
      <svg
        height={16}
        id="svg"
        width={16}
      />
      <canvas
        className="not-display"
        height={16}
        id="canvas"
        width={16}
      />
      <div>
        <a
          download="dots.png"
          href="#"
          id="download"
        >
          download
        </a>
      </div>
    </div>
  </div>
</div>
`;
